common/scala/src/main/scala/org/apache/openwhisk/core/connector/Message.scala

```diff
@@ -64,100 +64,165 @@ case class ActivationMessage(override val transid: TransactionId,
-                             activationId: ActivationId,
-                             isSystemError: Boolean,
-                             invoker: InvokerInstanceId)
+case class CombinedCompletionAndResultMessage(override val transid: TransactionId,
+                                              response: Either[ActivationId, WhiskActivation],
+                                              override val isSystemError: Option[Boolean],
+                                              invoker: InvokerInstanceId)
     extends AcknowledegmentMessage(transid) {
-
-  override def toString = {
-    activationId.asString
-  }
+  override def name = "combined"
+  override def result = Some(response)
+  override def isSlotFree = Some(invoker)
+  override def activationId = response.fold(identity, _.activationId)
+  override def toJson = CombinedCompletionAndResultMessage.serdes.write(this)
+  override def shrink = copy(response = response.flatMap(a => Left(a.activationId)))
+  override def toString = response.fold(identity, _.activationId).asString
 }
 
-object CompletionMessage extends DefaultJsonProtocol {
-  def parse(msg: String): Try[CompletionMessage] = Try(serdes.read(msg.parseJson))
-  implicit val serdes = jsonFormat4(CompletionMessage.apply)
+/**
+ * This message is sent from an invoker to the controller, once the resource slot in the invoker (used by the
+ * corresponding activation) free again (i.e., after log collection). The `CompletionMessage` is part of a split
+ * phase notification to the load balancer where an invoker first sends a `ResultMessage` and later sends the
+ * `CompletionMessage`.
+ */
+case class CompletionMessage(override val transid: TransactionId,
+                             override val activationId: ActivationId,
+                             override val isSystemError: Option[Boolean],
+                             invoker: InvokerInstanceId)
+    extends AcknowledegmentMessage(transid) {
+  override def name = "completion"
+  override def result = None
+  override def isSlotFree = Some(invoker)
+  override def toJson = CompletionMessage.serdes.write(this)
+  override def shrink = this
+  override def toString = activationId.asString
 }
 
 /**
- * That message will be sent from the invoker to the controller after action completion if the user wants to have
- * the result immediately (blocking activation).
- * When adding fields, the serdes of the companion object must be updated also.
- * The whisk activation field will have its logs stripped.
+ * This message is sent from an invoker to the load balancer once an action result is available for blocking actions.
+ * This is part of a split phase notification, and does not indicate that the slot is available, which is indicated with
+ * a `CompletionMessage`. Note that activation record will not contain any logs from the action execution, only the result.
  */
 case class ResultMessage(override val transid: TransactionId, response: Either[ActivationId, WhiskActivation])
```

Comment:
A result message should always contain a Right **except** when it must be shrunk because the message bus size limit on a message doesn't permit it to cross. So unless shrink is going to generate yet another type, I don't think we can remove the `Either` signature.

Reply:
Note that I pushed a change to tighten the constructors - you cannot cons a result with a Left - only a Right. To convert a Right to a Left you must use `shrink`.

Reply:
Thats the Right thing to do !!